### PR TITLE
Explore: updates ExploreItemState type, reduces strict null errors count to 702

### PR DESCRIPTION
**Pull request by Estrax**
**What this PR does / why we need it**:

~~Changes a couple of null values to undefined in Explore reducer. By doing so, it lowers the count of strict null errors from 726 to 701.~~

Updates ExploreItemState type - marks parameters that are always being initialized as non-optional.

### CODE REVIEW

**Comment by hugohaggmark** [comment]
Great work, for me undefined is used to signal the nonexisting/optional property in an object while null is used to signal that the value hasn't been initialized or reset. I would change the Explore state to be null instead of undefined in this case. WDYT @Estrax?

For instance, why does these need to be declared as an optional (undefined) property?
```ts
logsResult?: LogsModel;
graphResult?: GraphSeriesXY[];
tableResult?: DataFrame;
```

**Comment by Estrax** [comment]
@hugohaggmark good point. I also think having nulls there makes more sense (as it's not the variable definition stage but just a value change during the runtime), but somehow I assumed we follow the Microsoft's TypeScript style guide that clearly prefers undefined over null. I'll change the type definitions then. Thanks for a feedback!

**Comment by hugohaggmark** [comment]
> @hugohaggmark good point. I also think having nulls there makes more sense (as it's not the variable definition stage but just a value change during the runtime), but somehow I assumed we follow the Microsoft's TypeScript style guide that clearly prefers undefined over null. I'll change the type definitions then. Thanks for a feedback!

Didn't know about the Microsoft's TypeScript style guide, I need to do my home work it seems 🙈

**Comment by Estrax** [comment]
@hugohaggmark sure. There's a short guideline for development of TypeScript itself - that's only for their internal usage, not for other projects, but many projects follow the same style guide. I don't remember who, but someone brought it up during one of our calls: https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines

**Comment by hugohaggmark** [comment]
> @hugohaggmark sure. There's a short guideline for development of TypeScript itself - that's only for their internal usage, not for other projects, but many projects follow the same style guide. I don't remember who, but someone brought it up during one of our calls: https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines

Gotcha that was the one I looked at and as you say that's for the development of TypeScript itself. Thank you @Estrax

**Comment by aocenas** [comment]
Do we have to change nulls for undefined here? It seems that isn't the root of the type errors. I assume that means we assign nulls where where the type says it should be undefined is that right? Can we just change the type in that case to allow for null also? Not sure if that makes more sense here, but changing types so they are correct means 0 chance of introducing run time error while changing actual code from null to undefined can potentially cause trouble and I don't think we have enough test coverage to catch that.

**Comment by Estrax** [comment]
@aocenas yes, that's what we already talked with Hugo about. I'll revert those changes and make adjustments to the types instead.

**Comment by davkal** [comment]
> for me undefined is used to signal the nonexisting/optional property in an object while null is used to signal that the value hasn't been initialized or reset.

Same for me. I read those result types as values waiting to be filled, which will happen on query runs. There is nothing optional about them.

**Review by davkal — COMMENTED** [review]
The type changes work against my intuitions, so for me it's 👎 
But I'm also not married to them.

**Comment by davkal** [comment]
Why does this need the cherrypick label? It's not fixing important issues of the recent release.

**Comment by Estrax** [comment]
@davkal I feel we have a slight misunderstanding there (most likely because I wasn't clear enough with my intentions, sorry!). The type changes I am talking about are going to be made after reverting current changes from this PR. I'll check why do we have some of the variables (e.g. the ones mentioned by Hugo) set to be optional and if we keep them always defined (as there we set them to be null), then they no longer need to be optional, so then going to adjust the types accordingly.

**Comment by hugohaggmark** [comment]
Not my intention to start a lot of trouble for you @Estrax, I'm fine with whatever you think is correct for you.

**Comment by Estrax** [comment]
@hugohaggmark no need for sorry - it's all a valuable feedback! I'll push the changes soon.

**Comment by Estrax** [comment]
@hugohaggmark thanks again for your feedback and tips, fixed the type and got pretty much the same result, yet less invasive to the existing code.

File: public/app/features/explore/state/reducers.ts [hunk 1]
```diff
@@ -114,11 +115,16 @@ export const makeExploreItemState = (): ExploreItemState => ({
   update: makeInitialUpdateState(),
   latency: 0,
   supportedModes: [],
-  mode: null,
+  mode: (null as unknown) as ExploreMode,
```
Comment (aocenas) [inline public/app/features/explore/state/reducers.ts:118]: Shouldn't this be also `ExploreMode | null`? Same for `eventBridge`? Or is there some issue with that?

Comment (Estrax) [reply public/app/features/explore/state/reducers.ts]: Fixed. Just wasn't 100% sure with adding the null on those types as we don't set them to null anywhere else, so for the static type check to display an error if someone does that by mistake.

Comment (Estrax) [reply public/app/features/explore/state/reducers.ts:118]: @aocenas doubly checked - making those two variables possible to be nulls results in ~20 additional strict null errors. Not sure there's a point in fixing that now considering the `exploreMode` is going to be removed with the metrics & logs unification.

**Review by aocenas — APPROVED** [review]

**Pull request merged**